### PR TITLE
Update to mypy 0.991 for compatible-mypy & CI

### PR DESCRIPTION
**Pull request by intgr**

### CODE REVIEW

File: mypy.ini [hunk 1]
```diff
@@ -2,10 +2,11 @@
 strict_optional = True
 ignore_missing_imports = True
 check_untyped_defs = True
-warn_no_return = False
 show_traceback = True
 allow_redefinition = True
 incremental = True
+show_error_codes = False
+disable_error_code = empty-body
```
Comment (intgr) [inline mypy.ini:9]: `warn_no_return` is no longer necessary now that we disable `empty-body`.

File: rest_framework-stubs/permissions.pyi [hunk 2]
```diff
@@ -44,7 +44,11 @@ class NOT(_SupportsHasPermission):
     def __init__(self, op1: _SupportsHasPermission) -> None: ...
 
 class BasePermissionMetaclass(OperationHolderMixin, type): ...  # type: ignore[misc]
-class BasePermission(_SupportsHasPermission, metaclass=BasePermissionMetaclass): ...  # type: ignore[misc]
+
+class BasePermission(metaclass=BasePermissionMetaclass):  # type: ignore[misc]
+    def has_permission(self, request: Request, view: APIView) -> bool: ...
+    def has_object_permission(self, request: Request, view: APIView, obj: Any) -> bool: ...
```
Comment (intgr) [inline rest_framework-stubs/permissions.pyi:50]: If I kept `_SupportsHasPermission` here, it causes "Metaclass conflict" errors.

I don't understand metaclasses well enough to say whether that's a bug in mypy or a genuine issue.

But duplicating these method signatures here fixes it.

**Comment by intgr** [comment]
- [ ] typecheck-tests is not running correctly any more. All ignores show up in "UNUSED IGNORES" now.

File: scripts/typecheck_tests.py [hunk 3]
```diff
@@ -311,7 +316,8 @@ def update(self, op_code, cur_count, max_count=None, message=""):
         dir_util.copy_tree(str(STUBS_DIRECTORY), str(DRF_SOURCE_DIRECTORY / "rest_framework"))
     mypy_config_file = (PROJECT_DIRECTORY / "mypy.ini").absolute()
     mypy_cache_dir = PROJECT_DIRECTORY / ".mypy_cache"
-    tests_root = DRF_SOURCE_DIRECTORY / "tests"
+    # HACK: Work around https://github.com/python/mypy/issues/14042
+    tests_root = Path("drf_source") / "tests"
```
Comment (intgr) [inline scripts/typecheck_tests.py:320]: To be reverted

**Review by sobolevn — APPROVED** [review]
Thanks!

**Comment by intgr** [comment]
Oops @sobolevn I wanted to delay merging this to avoid creating a new conflict for #286. Probably should have left a comment about that. But hopefully this doesn't conflict.

**Pull request merged**